pkg/cmd/build/build.go

```diff
@@ -212,67 +212,172 @@ func translateDockerErr(err error) error {
 	return err
 }
 
-// OptsFromManifest returns the parsed options for the build from the manifest
-func OptsFromManifest(service string, b *model.BuildInfo, o *BuildOptions) *BuildOptions {
-	if b.Name == "" {
-		b.Name = os.Getenv(model.OktetoNameEnvVar)
+func getBuildOptionsFromManifest(service string, manifestBuildInfo *model.BuildInfo) *BuildOptions {
+	buildOptions := &BuildOptions{
+		Target:      manifestBuildInfo.Target,
+		Path:        manifestBuildInfo.Context,
+		CacheFrom:   manifestBuildInfo.CacheFrom,
+		ExportCache: manifestBuildInfo.ExportCache,
+		Tag:         manifestBuildInfo.Image,
+	}
+	if len(manifestBuildInfo.Args) > 0 {
+		buildOptions.BuildArgs = model.SerializeBuildArgs(manifestBuildInfo.Args)
 	}
 
-	args := model.SerializeBuildArgs(b.Args)
+	buildOptions.File = manifestBuildInfo.Dockerfile
+	if !filepath.IsAbs(manifestBuildInfo.Dockerfile) && !model.FileExistsAndNotDir(manifestBuildInfo.Dockerfile) {
+		buildOptions.File = filepath.Join(manifestBuildInfo.Context, manifestBuildInfo.Dockerfile)
+	}
 
-	if okteto.Context().IsOkteto && b.Image == "" {
-		tag := model.OktetoDefaultImageTag
+	if okteto.Context().IsOkteto && buildOptions.Tag == "" {
+		if manifestBuildInfo.Name == "" {
+			manifestBuildInfo.Name = os.Getenv(model.OktetoNameEnvVar)
+		}
 
 		envGitCommit := os.Getenv(model.OktetoGitCommitEnvVar)
 		isLocalEnvGitCommit := strings.HasPrefix(envGitCommit, model.OktetoGitCommitPrefix)
+		isPipeline := envGitCommit != "" && !isLocalEnvGitCommit
 
-		if envGitCommit != "" && !isLocalEnvGitCommit {
-			params := strings.Join(args, "") + envGitCommit
-			tag = fmt.Sprintf("%x", sha256.Sum256([]byte(params)))
-		}
-
-		// if flag --global, point to global registry
 		targetRegistry := okteto.DevRegistry
-		if o != nil && o.BuildToGlobal {
-			targetRegistry = okteto.GlobalRegistry
+		tag := model.OktetoDefaultImageTag
+
+		if len(manifestBuildInfo.VolumesToInclude) > 0 {
+			tag = model.OktetoImageTagWithVolumes
+		} else if isPipeline {
+			// if build is running at a pipeline with OKTETO_GIT_COMMIT, tag is replaced by a sha of this params
+			params := strings.Join(buildOptions.BuildArgs, "") + envGitCommit
+			tag = fmt.Sprintf("%x", sha256.Sum256([]byte(params)))
 		}
-		b.Image = fmt.Sprintf("%s/%s-%s:%s", targetRegistry, b.Name, service, tag)
-		if len(b.VolumesToInclude) > 0 {
-			b.Image = fmt.Sprintf("%s/%s-%s:%s", targetRegistry, b.Name, service, model.OktetoImageTagWithVolumes)
+		if manifestBuildInfo.Name != "" {
+			buildOptions.Tag = fmt.Sprintf("%s/%s-%s:%s", targetRegistry, manifestBuildInfo.Name, service, tag)
+		} else {
+			buildOptions.Tag = fmt.Sprintf("%s/%s:%s", targetRegistry, service, tag)
 		}
+	}
+
+	return buildOptions
+}
+
+func overrideManifestBuildOptions(manifestBuildOptions, cmdBuildOptions *BuildOptions) *BuildOptions {
```

Comment:
I love this functional approach. We should just be careful that we are working with pointers, and they point to the same underlying data that the parameters. Sadly go doesn't include any sort of deep copy to make this a `pure function`

Reply:
you can deep copy pointers doing something like this:

``` 
copyBuildOptions := &BuildOptions{}
*copyBuildOptions = *cmdBuildOptions
```

Reply:
The problem is that we have slices inside the struct that will not be copied by that

Reply:
There are some third party libraries that do this work of deep copying, some with code generation and others with `reflect`. 

Reading our `go.sum` seems like we have an indirect dependency to github.com/mohae/deepcopy, which seems like a simple solution. This one https://github.com/barkimedes/go-deepcopy on the other hand, is more robust since it takes into account recursive structs and that sort of pointer references